pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// GetNextTask returns the next Task for which a TaskRun should be created,
+// or nil if no TaskRun should be created.
+func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.SugaredLogger) *PipelineRunTaskRun {
+	for _, prtr := range state {
+		if prtr.TaskRun != nil {
+			c := prtr.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+			if c == nil {
+				logger.Infof("TaskRun %s doesn't have a condition so it is just starting and we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			}
+			switch c.Status {
+			case corev1.ConditionFalse:
+				logger.Infof("TaskRun %s has failed; we don't need to run PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			case corev1.ConditionUnknown:
+				logger.Infof("TaskRun %s is still running so we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			}
+		} else if canTaskRun(prtr.PipelineTask) {
+			logger.Infof("TaskRun %s should be started for PipelineRun %s", prtr.TaskRunName, prName)
+			return prtr
+		}
+	}
+	logger.Infof("No TaskRuns to start for PipelineRun %s", prName)
+	return nil
+}
+
+func canTaskRun(pt *v1alpha1.PipelineTask) bool {
+	// Check if Task can run now. Go through all the input constraints
+	return true
+}
+
+// PipelineRunTaskRun contains a Task and its associated TaskRun, if it
+// exists. TaskRun can be nil to represent there being no TaskRun.
+type PipelineRunTaskRun struct {
+	Task         *v1alpha1.Task
+	PipelineTask *v1alpha1.PipelineTask
+	TaskRunName  string
+	TaskRun      *v1alpha1.TaskRun
+}
+
+// GetTask is a function that will retrieve the Task name from namespace.
+type GetTask func(namespace, name string) (*v1alpha1.Task, error)
+
+// GetTaskRun is a function that will retrieve the TaskRun name from namespace.
+type GetTaskRun func(namespace, name string) (*v1alpha1.TaskRun, error)
```

Comment:
Not your fault, but this is not rendered properly :) i wonder if its a formatting issue

Reply:
yeah im not sure why the syntax highlighting is having a problem here :thinking: 

last time I saw this happen it was b/c @jonjohnsonjr decided to replace a `e` with an `ℯ` or something like that

Reply:
ah that could be it, thanks @jonjohnsonjr :D

and seriously that `𝛾` was pretty sweet, gonna be talking about that one for years to come

Reply:
https://github.com/knative/serving/pull/1460#discussion_r199993007

:rofl:

Reply:
![image](https://user-images.githubusercontent.com/432502/46836705-80bd1100-cd67-11e8-8118-7323cd38838a.png)

oh it just keeps getting better, i'm crying

Reply:
well that was distracing